plugins/SlicerT/SlicerTView.cpp

```diff
@@ -0,0 +1,192 @@
+#include "SlicerT.h"
+#include "Song.h"
+#include "StringPairDrag.h"
+#include "Track.h"
+#include "embed.h"
+
+namespace lmms {
+
+namespace gui {
+
+SlicerTView::SlicerTView(SlicerT* instrument, QWidget* parent)
+	: InstrumentViewFixedSize(instrument, parent)
+	, m_slicerTParent(instrument)
+{
+	// window settings
+	setAcceptDrops(true);
+	setAutoFillBackground(true);
+
+	// render background
+	QPalette pal;
+	pal.setBrush(backgroundRole(), PLUGIN_NAME::getIconPixmap("artwork"));
+	setPalette(pal);
+
+	m_wf = new SlicerTWaveform(248, 128, instrument, this);
+	m_wf->move(2, 6);
+
+	m_snapSetting = new ComboBox(this, tr("Slice snap"));
+	m_snapSetting->setGeometry(185, 200, 55, ComboBox::DEFAULT_HEIGHT);
+	m_snapSetting->setToolTip(tr("Set slice snapping for detection"));
+	m_snapSetting->setModel(&m_slicerTParent->m_sliceSnap);
+
+	m_syncToggle = new LedCheckBox("Sync", this, tr("SyncToggle"), LedCheckBox::LedColor::Green);
+	m_syncToggle->move(135, 187);
+	m_syncToggle->setToolTip(tr("Enable BPM sync"));
+	m_syncToggle->setModel(&m_slicerTParent->m_enableSync);
+
+	m_bpmBox = new LcdSpinBox(3, "19purple", this);
+	m_bpmBox->move(130, 201);
+	m_bpmBox->setToolTip(tr("Original sample BPM"));
+	m_bpmBox->setModel(&m_slicerTParent->m_originalBPM);
+
+	m_noteThresholdKnob = createStyledKnob();
+	m_noteThresholdKnob->move(10, 197);
+	m_noteThresholdKnob->setToolTip(tr("Threshold used for slicing"));
+	m_noteThresholdKnob->setModel(&m_slicerTParent->m_noteThreshold);
+
+	m_fadeOutKnob = createStyledKnob();
+	m_fadeOutKnob->move(64, 197);
+	m_fadeOutKnob->setToolTip(tr("Fade Out for notes"));
+	m_fadeOutKnob->setModel(&m_slicerTParent->m_fadeOutFrames);
+
+	m_midiExportButton = new QPushButton(this);
+	m_midiExportButton->move(199, 150);
+	m_midiExportButton->setIcon(PLUGIN_NAME::getIconPixmap("copyMidi"));
+	m_midiExportButton->setToolTip(tr("Copy midi pattern to clipboard"));
+	connect(m_midiExportButton, &PixmapButton::clicked, this, &SlicerTView::exportMidi);
+
+	m_resetButton = new QPushButton(this);
+	m_resetButton->move(18, 150);
+	m_resetButton->setIcon(PLUGIN_NAME::getIconPixmap("resetSlices"));
+	m_resetButton->setToolTip(tr("Reset Slices"));
+	connect(m_resetButton, &PixmapButton::clicked, m_slicerTParent, &SlicerT::updateSlices);
+}
+
+Knob* SlicerTView::createStyledKnob()
+{
+	Knob* newKnob = new Knob(KnobType::Styled, this);
+	newKnob->setFixedSize(50, 40);
+	newKnob->setCenterPointX(24.0);
+	newKnob->setCenterPointY(15.0);
+	return newKnob;
+}
+
+// copied from piano roll
+void SlicerTView::exportMidi()
+{
+	using namespace Clipboard;
+	if (m_slicerTParent->m_originalSample.frames() <= 1) { return; }
+
+	DataFile dataFile(DataFile::Type::ClipboardData);
+	QDomElement note_list = dataFile.createElement("note-list");
+	dataFile.content().appendChild(note_list);
+
+	std::vector<Note> notes = m_slicerTParent->getMidi();
+	if (notes.size() == 0) { return; }
+
+	TimePos start_pos(notes.front().pos().getBar(), 0);
+	for (Note note : notes)
```

Comment:
I believe you can (and should) take a reference here, i.e, use `Note&` instead of `Note`. It seems like you are only applying the changes to copies of the note, which will subsequently be discarded on the next iteration.

Reply:
Done